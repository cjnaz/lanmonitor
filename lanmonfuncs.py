#!/usr/bin/env python3
"""LAN monitor support functions
"""

# __version__ = "V1.0a 210515"

#==========================================================
#
#  Chris Nelson, 2021
#
# V1.1  210523  cmd timeout tweaks
# V1.0  210507  V1.0
# V1.0a 210515  Set timeouts to 1s for ping and 5s for ssh commands on remotes
#
# Changes pending
#   
#==========================================================

import sys
import os.path
import subprocess
import datetime
import time
import re
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../funcs3/'))
from funcs3 import logging, getcfg, snd_email, snd_notif

# Configs / Constants
NOTIF_SUBJ = "LAN Monitor"
RTN_PASS     = 0
RTN_WARNING  = 1
RTN_FAIL     = 2
RTN_CRITICAL = 3


# Module global vars
args = None
keylen = 0      # Padding vars used for pretty printing
hostlen = 0


def convert_time (timeX):
    """ Given time value term such as 90s, 15m, 20h, 3d, 2w return int seconds and units.
        <1h> returns (3600, "hours")
    If no units suffix, the value is interpreted as secs.
        <90> returns (90, "secs ")
    """
    if type(timeX) is int:
        return timeX, "secs "           # Case Int
    try:
        return int(timeX), "secs "      # Case Str without units
    except:
        pass
    try:                                # Case Str with units
        time_value = int(timeX[:-1])
        time_unit =  timeX[-1:].lower()
        if time_unit == "s":
            return time_value, "secs "
        if time_unit == "m":
            return time_value * 60, "mins "
        if time_unit == "h":
            return time_value * 60*60, "hours"
        if time_unit == "d":
            return time_value * 60*60*24, "days "
        if time_unit == "w":
            return time_value * 60*60*24*7, "weeks"
        raise ValueError(f"Illegal time units <{time_unit}>")
    except Exception as e:
        logging.error (f"ERROR:  Could not convert time value <{timeX}>\n  {e}")
        raise


USER_HOST_FORMAT = re.compile(r"[\w.-]+@([\w.-]+)$")
USER_HOST_PORT_FORMAT = re.compile(r"[\w.-]+@([\w.-]+):([\d]+)$")

def split_user_host_port(u_h_p):
    """ Handle variations in passed-in user_host_port (u_h_p)
        "local", "user@host", or "user@host:port"
    Return separate user_host, host, and port values
    EG:
        "xyz@host15:4455" returns:
            ["xyz@host15", "host15", "4455"]
        "xyz@host15" returns:
            ["xyz@host15", "host15", "22"]  (the default port is "22" for ssh)
        "local" returns
            ["local", "local", ""]
    """
    user_host_noport = u_h_p.split(":")[0]
    host = u_h_p
    port = ""
    if host != "local":
        port = "22"
        out = USER_HOST_FORMAT.match(u_h_p)
        if out:
            host = out.group(1)
        else:
            out = USER_HOST_PORT_FORMAT.match(u_h_p)
            if out:
                host = out.group(1)
                port = out.group(2)
            else:
                _msg = f"Expecting <user@host> or <user@host:port> format, but found <{u_h_p}>."
                logging.error(f"ERROR:  {_msg}")
                raise ValueError (_msg)
    return user_host_noport, host, port


def cmd_check(cmd, user_host_port, return_type=None, check_line_text=None, expected_text=None, not_text=None):
    """  Runs the cmd and operates on the response based on return_type.
    return_types:
        check_string        Returns True if expected_text occurs in response of the cmd, plus the full full subprocess run structure
            check_line_text     If provided, only the first line containing this text is checked
            expected_text       Text that must be found
            not_text            Text that must NOT be found
        cmdrun              Returns True if the cmd return code was 0, plus the full full subprocess run structure
    """

    if user_host_port != "local":
        u_h, _, port = split_user_host_port(user_host_port)
        cmd = ["ssh", u_h, "-p" + port, "-o", "ConnectTimeout=1", "-T"] + cmd

    for nTry in range (getcfg('nRetries')):
        try:
            logging.debug(f"cmd_check subprocess command try {nTry+1}: <{cmd}>")
            # runtry = subprocess.run(cmd, capture_output=True, text=True)  # Py 3.7+
            runtry = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)   #Py3.6 requires old-style params
        except Exception as e:
            logging.error(f"ERROR:  subprocess.run of cmd <{cmd}> failed.\n  {e}")
            return False, None

        if return_type == "check_string":
            if check_line_text is None:
                text_to_check = runtry.stdout
            else:
                text_to_check = ""
                for line in runtry.stdout.split("\n"):
                    if check_line_text in line:
                        text_to_check = line
                        break
                
            if expected_text in text_to_check:
                if not_text is not None:
                    if not_text not in text_to_check:
                        return True, runtry
                else:
                    return True, runtry
            else:
                if nTry == getcfg('nRetries')-1:
                    return False, runtry

        elif return_type == "cmdrun":
            if runtry.returncode == 0:
                return True, runtry
            elif nTry == getcfg('nRetries')-1:
                return False, runtry

        else:
            _msg = f"Invalid return_type <{return_type}> passed to cmd_check"
            logging.error (f"ERROR:  {_msg}")
            raise ValueError (_msg)
        time.sleep (convert_time(getcfg('RetryInterval'))[0])


have_access_dict = {}

def have_access(user_host_port):
    """If user_host_port != "local" then check ping and ssh access to the host.
    Remember passing result and avoid redoing the ping and ssh tests on later checks.
    If ping/ssh tests fail they will be retried if referenced in a later monitored item.
    The dictionary is cleared on each RecheckInterval in the lanmonitor main code.
    ping or ssh access fails return a tuple of (False, error_message_string).
    (True, "") is returned if local or if ping and ssh tests pass for non-local host.

    NOTE that this optimization can result in confusion:
        if access once passed then fails the plugin will be called and may flag an error
        while the actual problem is no access.
    """
    user_host_port = user_host_port.lower()
    if user_host_port == "local":
        return (True, "")
    if user_host_port in have_access_dict:
        return (True, "")

    _, host, _ = split_user_host_port(user_host_port)
    if host != "local":     # Check for remote access if non-local
        pingrslt = cmd_check(["ping", "-c", "1", "-W", "1", host], user_host_port="local", return_type="cmdrun")
        if not pingrslt[0]:
            return (False, "HOST CANNOT BE REACHED")
        else:
            sshrslt = cmd_check(["echo", "hello"], user_host_port=user_host_port, return_type="cmdrun")
            if "Connection refused" in sshrslt[1].stderr:
                return (False, f"SSH CONNECTION REFUSED <{user_host_port}>")
            # if "The authenticity of host" in sshrslt[1].stdout:       # Never happens thru the script, only interactive
            #     return (False, f"SSH CONNECTION NOT APPROVED.  ssh to <{user_host_port}> manually first.")

    have_access_dict[user_host_port] = True
    return (True, "")


def next_summary_timestring():
    """Calculate timestring of next summary.
    SummaryDays			0 1 2 3 4 5 6	# Days of week.  0 = Sunday
    SummaryTime			9:45		    # 24 hour clock
    NOTE:  May be off by 1 hour over a DTS changes.
    """
    try:
        target_hour   = int(getcfg("SummaryTime","").split(":")[0])
        target_minute = int(getcfg("SummaryTime","").split(":")[1])
        today_day_num = datetime.datetime.today().isoweekday()    # Sunday = 0, Saturday = 6
        now = datetime.datetime.now().replace(second=0, microsecond=0)
        next_summary = now + datetime.timedelta(days=30)
        for daynum in getcfg("SummaryDays").split():
            offset_num_days = int(daynum) - today_day_num
            if offset_num_days < 0:
                offset_num_days += 7
            plus_day =  now + datetime.timedelta(days=offset_num_days)
            with_time = plus_day.replace(hour=target_hour, minute=target_minute)
            if with_time < datetime.datetime.now():
                with_time = with_time + datetime.timedelta(days=7)
            if with_time < next_summary:
                next_summary = with_time
        logging.debug(f"Next summary:  {next_summary}")
        return next_summary
    except Exception as e:
        _msg = f"SummaryDays <{getcfg('SummaryDays','')}> or SummaryTime <{getcfg('SummaryTime','')}> settings could not be parsed\n  {e}"
        logging.error(f"ERROR:  {_msg}")
        raise ValueError (_msg) from None
        