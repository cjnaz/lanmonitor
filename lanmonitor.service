[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path to>/lanmonitor

[Install]
WantedBy=multi-user.target
