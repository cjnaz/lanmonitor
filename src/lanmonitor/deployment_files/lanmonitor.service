[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path_to>/lanmonitor --service

[Install]
WantedBy=multi-user.target
