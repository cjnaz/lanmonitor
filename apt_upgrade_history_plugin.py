#!/usr/bin/env python3
"""LAN Monitor plugin - apt_upgrade_history_plugin

apt history files at /var/log/apt/* are checked for the specific <apt_command> text and the date is extracted from
the most recent occurrence of this text.  May require root access in order to read apt history.

Typical config file lines:
    MonType_AptUpgrade  apt_upgrade_history_plugin
    AptUpgrade_<friendly_name>  <local or user@host>  [CRITICAL]  <age>  <apt_command>
    AptUpgrade_MyHost  local  CRITICAL  15d  apt full-upgrade
"""

__version__ = "V1.0 221120"

#==========================================================
#
#  Chris Nelson, 2021-2022
#
# V1.0  221120  Initial
#
# Changes pending
#  The apt history file content may be locale specific.  This plugin is currently hardcoded to US style.
#   
#==========================================================

import datetime
import re
import globvars
from lanmonfuncs import RTN_PASS, RTN_WARNING, RTN_FAIL, RTN_CRITICAL, cmd_check #, convert_time
from funcs3 import logging, timevalue, retime

# Configs / Constants
HISTORY_FILES = r"/var/log/apt/history.log*"
APTLINEFORMAT = r"Start-Date: ([\d-]+  [\d:]+)\sCommandline: "
# Start-Date: 2022-10-11  20:09:49
# Commandline: apt full-upgrade


class monitor:

    def __init__ (self):
        pass

    def setup (self, item):
        """ Set up instance vars and check item values.
        Passed in item dictionary keys:
            key             Full 'itemtype_tag' key value from config file line
            tag             'tag' portion only from 'itemtype_tag' from config file line
            user_host_port  'local' or 'user@hostname[:port]' from config file line
            host            'local' or 'hostname' from config file line
            critical        True if 'CRITICAL' is in the config file line
            rest_of_line    Remainder of line after the 'user_host' from the config file line
        Returns True if all good, else False
        """

        # Construct item type specifics and check validity
        self.key            = item["key"]                           # vvvv These items don't need to be modified
        self.key_padded     = self.key.ljust(globvars.keylen)
        self.tag            = item["tag"]
        self.user_host_port = item["user_host_port"]
        self.host           = item["host"]
        self.host_padded    = self.host.ljust(globvars.hostlen)
        if item["critical"]:
            self.failtype = RTN_CRITICAL
            self.failtext = "CRITICAL"
        else:
            self.failtype = RTN_FAIL
            self.failtext = "FAIL"                                  # ^^^^ These items don't need to be modified

        try:
            xx = item["rest_of_line"].split(maxsplit=1)
            maxagevar = timevalue(xx[0])
            self.maxage_sec = maxagevar.seconds
            self.units = maxagevar.unit_str
            self.unitsC = maxagevar.unit_char
            self.apt_command = xx[1]
            self.apt_RE = re.compile(APTLINEFORMAT + self.apt_command + "\n")
        except Exception as e:
            logging.error (f"  ERROR:  <{self.key}> INVALID LINE SYNTAX <{item['rest_of_line']}>\n  {e}")
            return RTN_FAIL

        return RTN_PASS


    def eval_status (self):
        """ Check status of this item.
        Returns dictionary with these keys:
            rslt            Integer status:  RTN_PASS, RTN_WARNING, RTN_FAIL, RTN_CRITICAL
            notif_key       Unique handle for tracking active notifications in the notification handler 
            message         String with status and context details
        """
        
        cmd = ["zcat", "-qf", "/var/log/apt/history.log*"]
        rslt = cmd_check(cmd, user_host_port=self.user_host_port, return_type="cmdrun")
        # print (rslt)                  # Uncomment for debug

        if not rslt[0]:
            return {"rslt":RTN_WARNING, "notif_key":self.key, "message":f"  WARNING: {self.key} - {self.host} - COULD NOT GET APT HISTORY"}

        dates = self.apt_RE.findall(rslt[1].stdout)
        if len(dates) == 0:
            return {"rslt":RTN_WARNING, "notif_key":self.key, "message":f"  WARNING: {self.key} - {self.host} - NO <{self.apt_command}> COMMANDS IN THE APT HISTORY LOGS"}

        last_update = 0
        for date in dates:
            timestamp = datetime.datetime.strptime(date, "%Y-%m-%d  %H:%M:%S").timestamp()
            if timestamp > last_update:
                last_update = timestamp
        last_update_age = datetime.datetime.now().timestamp() - last_update
        if last_update_age < self.maxage_sec:
            return {"rslt":RTN_PASS, "notif_key":self.key, "message":f"{self.key_padded}  OK - {self.host_padded} - {retime(last_update_age, self.unitsC):6.1f} {self.units:5} ({int(retime(self.maxage_sec, self.unitsC)):>4} {self.units:5} max)"}
        else:
            return {"rslt":self.failtype, "notif_key":self.key, "message":f"  {self.failtext}: {self.key} - {self.host} - APT UPGRADE TOO LONG AGO - {retime(last_update_age, self.unitsC):6.1f} {self.units:5} ({int(retime(self.maxage_sec, self.unitsC)):>4} {self.units:5} max)"}


if __name__ == '__main__':
    import argparse
    from funcs3 import loadconfig

    CONFIG_FILE = "lanmonitor.cfg"

    parser = argparse.ArgumentParser(description=__doc__ + __version__, formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('--config-file', default=CONFIG_FILE,
                        help=f"Path to config file (default <{CONFIG_FILE}>).")
    parser.add_argument('-V', '--version', action='version', version='%(prog)s ' + __version__,
                        help="Return version number and exit.")

    globvars.args = parser.parse_args()
    loadconfig(cfgfile=globvars.args.config_file)


    def dotest (test):
        print (f"\n{test}")
        inst = monitor()
        setup_rslt = inst.setup(test)
        print (f"  {setup_rslt}")
        if setup_rslt == RTN_PASS:
            print(f"  {inst.eval_status()}")

    dotest ({"key":"AptUpgrade_Pass", "tag":"Pass", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "rest_of_line":"500w apt full-upgrade"})

    dotest ({"key":"AptUpgrade_TooOld", "tag":"TooOld", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "rest_of_line":"10h apt full-upgrade"})

    dotest ({"key":"AptUpgrade_NoUpgrades", "tag":"NoUpgrades", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "rest_of_line":"10h apt full-upgradeS"})

    dotest ({"key":"AptUpgrade_CantAccess", "tag":"CantAccess", "host":"nosuchhost", "user_host_port":"pi@nosuchhost", "critical":True, "rest_of_line":"10h apt full-upgrade"})

    dotest ({"key":"AptUpgrade_baddef", "tag":"badline", "host":"local", "user_host_port":"local", "critical":True, "rest_of_line":"10m"})

    dotest ({"key":"AptUpgrade_badtime", "tag":"badtime", "host":"local", "user_host_port":"local", "critical":True, "rest_of_line":"10y apt full-upgrade"})

