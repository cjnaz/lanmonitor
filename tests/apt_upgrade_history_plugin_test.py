#!/usr/bin/env python3
"""LAN Monitor plugin - apt_upgrade_history_plugin_test
"""

#==========================================================
#
#  Chris Nelson, Copyright 2021-2023
#
# 3.0 230301 - Packaged
#
#==========================================================

from cjnfuncs.cjnfuncs import set_toolname, setuplogging, logging, cfg
from lanmonitor.lanmonfuncs import RTN_PASS
try:
    from apt_upgrade_history_plugin import monitor
except:
    from lanmonitor.apt_upgrade_history_plugin import monitor

tool = set_toolname("tool")
cfg["nRetries"]         = 1
cfg["RetryInterval"]    = "0s"
cfg["ConsoleLogFormat"] = "{module:>35}.{funcName:20} - {levelname:>8}:  {message}"
setuplogging()
logging.getLogger().setLevel(logging.DEBUG)

test_num = 0
def dotest (test):
    global test_num
    test_num += 1
    print(f"\nTest {test_num} {'-'*50}")
    inst = monitor()
    setup_rslt = inst.setup(test)
    logging.debug (f"{test['key']} - setup() returned:  {setup_rslt}")
    if setup_rslt == RTN_PASS:
        logging.debug (f"{test['key']} - eval_status() returned:  {inst.eval_status()}")

dotest ({"key":"AptUpgrade_Pass", "tag":"Pass", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"500w apt full-upgrade"})

dotest ({"key":"AptUpgrade_TooOld", "tag":"TooOld", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"10h apt full-upgrade"})

dotest ({"key":"AptUpgrade_NoUpgrades", "tag":"NoUpgrades", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"10h apt full-upgradeS"})

dotest ({"key":"AptUpgrade_CantAccess", "tag":"CantAccess", "host":"nosuchhost", "user_host_port":"pi@nosuchhost", "critical":True, "check_interval":1, "rest_of_line":"10h apt full-upgrade"})

dotest ({"key":"AptUpgrade_baddef", "tag":"badline", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10m"})

dotest ({"key":"AptUpgrade_badtime", "tag":"badtime", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"10y apt full-upgrade"})
