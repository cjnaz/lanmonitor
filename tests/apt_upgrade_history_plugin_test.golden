
Test 1 - Local pass - OK --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_Pass - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_Pass', 'tag': 'Pass', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '500w apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Pass - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_Pass - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['zcat', '-qf', '/var/log/apt/history.log.5.gz', '/var/log/apt/history.log.3.gz', '/var/log/apt/history.log.1.gz', '/var/log/apt/history.log.2.gz', '/var/log/apt/history.log.4.gz', '/var/log/apt/history.log']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'AptUpgrade_Pass', 'message': 'AptUpgrade_Pass  OK - local -   14.0 weeks ( 500 weeks max)'}

Test 2 - Local too old - CRITICAL --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_TooOld - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_TooOld', 'tag': 'TooOld', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1m apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_TooOld - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_TooOld - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['zcat', '-qf', '/var/log/apt/history.log.5.gz', '/var/log/apt/history.log.3.gz', '/var/log/apt/history.log.1.gz', '/var/log/apt/history.log.2.gz', '/var/log/apt/history.log.4.gz', '/var/log/apt/history.log']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_TooOld - eval_status() returned:  {'rslt': 3, 'notif_key': 'AptUpgrade_TooOld', 'message': '  CRITICAL: AptUpgrade_TooOld - local - APT UPGRADE TOO LONG AGO - 140810.1 mins  (   1 mins  max)'}

Test 3 - Local no Upgrades in history - WARNING --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_NoUpgrades - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_NoUpgrades', 'tag': 'NoUpgrades', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10h apt full-upgradeS'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_NoUpgrades - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_NoUpgrades - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['zcat', '-qf', '/var/log/apt/history.log.5.gz', '/var/log/apt/history.log.3.gz', '/var/log/apt/history.log.1.gz', '/var/log/apt/history.log.2.gz', '/var/log/apt/history.log.4.gz', '/var/log/apt/history.log']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_NoUpgrades - eval_status() returned:  {'rslt': 1, 'notif_key': 'AptUpgrade_NoUpgrades', 'message': '  WARNING: AptUpgrade_NoUpgrades - local - NO <apt full-upgradeS> COMMANDS IN THE APT HISTORY LOGS'}

Test 4 - Remote pass - OK --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_RemotePass - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_RemotePass', 'tag': 'RemotePass', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '500w apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemotePass - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_RemotePass - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemotePass - eval_status() returned:  {'rslt': 0, 'notif_key': 'AptUpgrade_RemotePass', 'message': 'AptUpgrade_RemotePass  OK - testhost2 -   14.0 weeks ( 500 weeks max)'}

Test 5 - Remote too old - CRITICAL --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_RemoteTooOld - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_RemoteTooOld', 'tag': 'RemoteTooOld', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1m apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemoteTooOld - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_RemoteTooOld - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemoteTooOld - eval_status() returned:  {'rslt': 3, 'notif_key': 'AptUpgrade_RemoteTooOld', 'message': '  CRITICAL: AptUpgrade_RemoteTooOld - testhost2 - APT UPGRADE TOO LONG AGO - 140810.2 mins  (   1 mins  max)'}

Test 6 - Remote no Upgrades in history - WARNING --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_RemoteNoUpgrades - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_RemoteNoUpgrades', 'tag': 'RemoteNoUpgrades', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10h apt full-upgradeS'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemoteNoUpgrades - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_RemoteNoUpgrades - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_RemoteNoUpgrades - eval_status() returned:  {'rslt': 1, 'notif_key': 'AptUpgrade_RemoteNoUpgrades', 'message': '  WARNING: AptUpgrade_RemoteNoUpgrades - testhost2 - NO <apt full-upgradeS> COMMANDS IN THE APT HISTORY LOGS'}

Test 7 - No such host - WARNING --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_Unknown - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_Unknown', 'tag': 'Unknown', 'host': 'nosuchhost', 'user_host_port': 'me@nosuchhost', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10h apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Unknown - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_Unknown - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: Could not resolve hostname nosuchhost: Name or service not known> - attempting simplessh connection to <me@nosuchhost>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: Could not resolve hostname nosuchhost: Name or service not known>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Unknown - eval_status() returned:  {'rslt': 1, 'notif_key': 'AptUpgrade_Unknown', 'message': '  WARNING: AptUpgrade_Unknown - nosuchhost - ssh: Could not resolve hostname nosuchhost: Name or service not known'}

Test 8 - Known host, unavailable - WARNING --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_Unavailable - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_Unavailable', 'tag': 'Unavailable', 'host': 'testhostX', 'user_host_port': 'me@testhostX', 'critical': True, 'cmd_timeout': 4, 'check_interval': 1, 'rest_of_line': '10h apt full-upgrade'}
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Unavailable - setup() returned:  0
         apt_upgrade_history_plugin.eval_status          -    DEBUG:  AptUpgrade_Unavailable - lanmonitor.apt_upgrade_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@testhostX', '-p22', '-T', 'zcat', '-qf', '/var/log/apt/history.log*']>:  ssh: connect to host testhostx port 22: No route to host
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: connect to host testhostx port 22: No route to host> - attempting simplessh connection to <me@testhostX>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>:  ssh: connect to host testhostx port 22: No route to host
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: connect to host testhostx port 22: No route to host>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_Unavailable - eval_status() returned:  {'rslt': 1, 'notif_key': 'AptUpgrade_Unavailable', 'message': '  WARNING: AptUpgrade_Unavailable - testhostX - ssh: connect to host testhostx port 22: No route to host'}

Test 9 - Bad (no) check command - setup ERROR --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_baddef - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_baddef', 'tag': 'badline', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10m'}
         apt_upgrade_history_plugin.setup                -    ERROR:    ERROR:  <AptUpgrade_baddef> INVALID LINE SYNTAX <10m>
  list index out of range
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_baddef - setup() returned:  2

Test 10 - Bad time limit - setup ERROR --------------------------------------------------
         apt_upgrade_history_plugin.setup                -    DEBUG:  AptUpgrade_badtime - lanmonitor.apt_upgrade_history_plugin.setup() called:
  {'key': 'AptUpgrade_badtime', 'tag': 'badtime', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10y apt full-upgrade'}
         apt_upgrade_history_plugin.setup                -    ERROR:    ERROR:  <AptUpgrade_badtime> INVALID LINE SYNTAX <10y apt full-upgrade>
  Illegal time units <y> in time string <10y>
    apt_upgrade_history_plugin_test.dotest               -    DEBUG:  AptUpgrade_badtime - setup() returned:  2
